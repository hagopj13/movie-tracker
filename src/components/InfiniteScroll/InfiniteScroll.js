// @flow
import React, { useEffect, useCallback } from 'react';
import type { Node } from 'react';

type Props = {
  children: Node,
  activationDistance?: number,
  onLoadMore: () => void,
};

const InfiniteScroll = (props: Props) => {
  const { children, activationDistance, onLoadMore } = props;

  const handleScroll = useCallback(() => {
    const documentHeight = document.body?.offsetHeight ?? 0;
    const scrollHeight = window.innerHeight + window.scrollY;
    if (scrollHeight + activationDistance >= documentHeight) {
      onLoadMore();
    }
  }, [activationDistance, onLoadMore]);

  useEffect(() => {
    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, [handleScroll]);

  return <>{children}</>;
};

InfiniteScroll.defaultProps = {
  activationDistance: 20,
};

export default InfiniteScroll;
