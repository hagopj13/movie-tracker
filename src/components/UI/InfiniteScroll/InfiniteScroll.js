// @flow
import React, { useEffect, useCallback } from 'react';
import type { Node } from 'react';

type Props = {
  children: Node,
  onLoadMore: () => void,
};

const InfiniteScroll = (props: Props) => {
  const { children, onLoadMore } = props;

  const activationDistance = 20;
  const handleScroll = useCallback(() => {
    const documentHeight = document.body?.offsetHeight;
    const scrollHeight = window.innerHeight + window.scrollY;
    if (scrollHeight + activationDistance >= documentHeight) {
      onLoadMore();
    }
  }, [onLoadMore]);

  useEffect(() => {
    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, [handleScroll]);

  return <>{children}</>;
};

export default InfiniteScroll;
