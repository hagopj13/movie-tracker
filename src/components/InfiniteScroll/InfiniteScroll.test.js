import React from 'react';
import { createMount } from '@material-ui/core/test-utils';

import InfiniteScroll from './InfiniteScroll';

describe('InfiniteScroll component', () => {
  let mount;
  let eventsMap;
  let wrapper;

  const mockScrollCallbackRemove = jest.fn();
  const mockLoadMore = jest.fn();
  const ChildComponent = () => <div>Child component</div>;

  beforeAll(() => {
    mount = createMount();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    eventsMap = new Map();
    global.addEventListener = jest.fn().mockImplementation((event, callback) => {
      eventsMap.set(event, callback);
    });
    global.removeEventListener = jest.fn().mockImplementation((event) => {
      if (event === 'scroll') {
        mockScrollCallbackRemove();
      }
    });
    const mockProps = {
      activationDistance: 0,
      onLoadMore: mockLoadMore,
    };
    wrapper = mount(
      <InfiniteScroll {...mockProps}>
        <ChildComponent />
      </InfiniteScroll>,
    );
  });

  it('should render the InfiniteScroll component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should fire onLoadMore callback when scroll happens and reaches the bottom', () => {
    expect(mockLoadMore).not.toHaveBeenCalled();
    eventsMap.get('scroll')();
    expect(mockLoadMore).toHaveBeenCalledTimes(1);
  });

  it('should not fire onLoadMore callback after scroll happens but does not reach the bottom', () => {
    expect(mockLoadMore).not.toHaveBeenCalled();
    global.innerHeight = -1;
    global.scrollY = 0;
    eventsMap.get('scroll')();
    expect(mockLoadMore).not.toHaveBeenCalled();
  });

  it('should remove the scroll listener when unmounting', () => {
    // Because of how useEffect works, the scroll listener could have already been removed a few times
    const scrollListenerRemovedTimes = mockScrollCallbackRemove.mock.calls.length;
    wrapper.unmount();
    expect(mockScrollCallbackRemove).toHaveBeenCalledTimes(scrollListenerRemovedTimes + 1);
  });
});
