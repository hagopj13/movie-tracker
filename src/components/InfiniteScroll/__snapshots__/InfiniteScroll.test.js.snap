// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll component should render the InfiniteScroll component correctly 1`] = `
<InfiniteScroll
  activationDistance={0}
  onLoadMore={[MockFunction]}
>
  <ChildComponent>
    <div>
      Child component
    </div>
  </ChildComponent>
</InfiniteScroll>
`;
